[mlir][linalg] Prevent hoisting of transfer pairs in the presence of aliases

banach-space opened the pull request:
This patch adds additional checks to the hoisting logic to prevent hoisting of
`vector.transfer_read` / `vector.transfer_write` pairs when the underlying memref
has users that introduce aliases via operations implementing
`ViewLikeOpInterface`.

Note: This may conservatively block some valid hoisting opportunities and could
affect performance. However, as demonstrated by the included tests, the current
logic is too permissive and can lead to incorrect transformations.

If this change prevents hoisting in cases that are provably safe, please share
a minimal repro — I'm happy to explore ways to relax the check.

Special treatment is given to `memref.assume_alignment`, mainly to accommodate
recent updates in:

* https://github.com/llvm/llvm-project/pull/139521

Note that such special casing does not scale and should generally be avoided.
The current hoisting logic lacks robust alias analysis. While better support
would require more work, the broader semantics of `memref.assume_alignment`
remain somewhat unclear. It's possible this op may eventually be replaced with
the "alignment" attribute added in:

* https://github.com/llvm/llvm-project/pull/144344

banach-space committed:
[mlir][linalg] Prevent hoisting of transfer pairs in the presence of aliases

This patch adds additional checks to the hoisting logic to prevent
hoisting of `vector.transfer_read`/`vector.transfer_write` pairs when
the underlying `memref` has users that introduce aliases via operations
implementing `ViewLikeOpInterface`.

Note: This may conservatively block some valid hoisting opportunities
and could impact performance. However, as demonstrated by the included
tests, the current behavior is too permissive and can lead to incorrect
transformations.

If this change prevents hoisting in cases that are provably safe, please
share a minimal repro — I’d be happy to explore ways to relax the check.

xiangzh1 commented on mlir/lib/Dialect/Linalg/Transforms/Hoisting.cpp:
```diff
@@ -312,14 +312,27 @@ void mlir::linalg::hoistRedundantVectorTransfers(Operation *root,
 
+      auto base = transferRead.getBase();
+      for (auto *user : base.getUsers())
+        if (isa_and_nonnull<ViewLikeOpInterface>(user))
```
Seems better to check the confilction at  
[Hoisting.cpp:Line349](https://github.com/llvm/llvm-project/blob/f7218d71dd69f63b14e6fb5d3da06228982b7423/mlir/lib/Dialect/Linalg/Transforms/Hoisting.cpp#L349) than cancel the optimization.

banach-space committed:
fixup! [mlir][linalg] Prevent hoisting of transfer pairs in the presence of aliases

1. Relax the conditions in the case of `memref.assume_alignment`. This
   unblocks hoisting for examples like the one reported in
   https://github.com/llvm/llvm-project/issues/144825 (see the link in
   the top post: "detailed example pls refer to example").
2. When checking the source operand for the xfer Ops, we only need to
   look at either xfer_write or xfer_read (we already know that the
   source is identical). The corresponding logic has been simplified.

xiangzh1 commented on mlir/lib/Dialect/Linalg/Transforms/Hoisting.cpp:
```diff
@@ -303,23 +303,45 @@ void mlir::linalg::hoistRedundantVectorTransfers(Operation *root,
+        //  `memref.assume_alignment`)
+        if (auto assume = dyn_cast<memref::AssumeAlignmentOp>(source)) {
+          Value memPreAlignment = assume.getMemref();
+          if (base.hasNUses(2) && memPreAlignment.hasOneUse())
```
We should check the use num inside the loop, due to the base and memPreAlignment are usually used outside of the loop which should not affect the optimization.

banach-space committed:
fixup! fixup! [mlir][linalg] Prevent hoisting of transfer pairs in the presence of aliases

Make sure only uses inside the loop are counted

xiangzh1 reviewed (approved):
LGTM， thanks for your effect.

hanhanW commented on mlir/lib/Dialect/Linalg/Transforms/Hoisting.cpp:
```diff
@@ -303,23 +303,50 @@ void mlir::linalg::hoistRedundantVectorTransfers(Operation *root,
+      auto base = transferRead.getBase();
+      auto *source = base.getDefiningOp();
+      if (source) {
+        // NOTE: We treat `memref.assume_alignment` as a special case:
+        //  1. If it has exactly two uses then these have to be the xfer Ops
+        //    being looked at.
+        //  2. Otherwise, there are other users that we should take into
+        //    account
+        // In the case of 1., it is safe to look past AssumeAlignmentOp,
+        // i.e. at the defining Op of the input MemRef, provided that:
+        //  * the original MemRef has only one use (i.e.
+        //  `memref.assume_alignment`)
```
nit: format it a bit, e.g., alignment, missing period at the end.

```suggestion
        // NOTE: We treat `memref.assume_alignment` as a special case:
        //  1. If it has exactly two uses then these have to be the xfer Ops
        //     being looked at.
        //  2. Otherwise, there are other users that we should take into
        //     account.
        // In the case of 1., it is safe to look past AssumeAlignmentOp,
        // i.e. at the defining Op of the input MemRef, provided that:
        //  * the original MemRef has only one use (i.e.
        //    `memref.assume_alignment`).
```

hanhanW commented on mlir/lib/Dialect/Linalg/Transforms/Hoisting.cpp:
```diff
@@ -303,23 +303,50 @@ void mlir::linalg::hoistRedundantVectorTransfers(Operation *root,
-      source = transferWrite.getBase().getDefiningOp();
-      if (source && isa_and_nonnull<ViewLikeOpInterface>(source))
-        return WalkResult::advance();
+      for (auto *user : base.getUsers())
+        if (isa_and_nonnull<ViewLikeOpInterface>(user))
+          return WalkResult::advance();
```
optional nit: I'd use any_of and `llvm::IsaPred<>` in this case. (I think the user won't be NULL, so dropping nonnull should be okay.) E.g.,

```cpp
if (llvm::any_of(base.getUsers(), llvm::IsaPred<ViewLikeOpInterface>))
  return WalkResult::advance();
```

hanhanW reviewed (approved):
Thanks, it looks better!

banach-space committed:
fixup! fixup! fixup! [mlir][linalg] Prevent hoisting of transfer pairs in the presence of aliases

Incorporate suggestions from HanHan, update the description.

banach-space committed:
fixup! fixup! fixup! fixup! [mlir][linalg] Prevent hoisting of transfer pairs in the presence of aliases

Extra test

banach-space commented:
**UPDATES (26/6/25):**

* Rebased on top of `main` (to incorporate https://github.com/llvm/llvm-project/pull/145234 and to fix merge conflicts)
* Re-wrote the comment in Hoisting.cpp that explains the rationale for special-casing `memref.assume_alignment`.
* Added a negative test with `memref.assume_alignment`.

If there are no new comments, I will land it before the end of the week. Thanks for reviews!
